agent/structs/config_entry.go

```diff
@@ -159,11 +162,83 @@ func (e *ProxyConfigEntry) GetRaftIndex() *RaftIndex {
 	return &e.RaftIndex
 }
 
+func (c *ProxyConfigEntry) MarshalBinary() (data []byte, err error) {
+	type alias ProxyConfigEntry
+
+	a := alias(*c)
+
+	// bs will grow if needed but allocate enough to avoid reallocation in common
+	// case.
+	bs := make([]byte, 128)
+	enc := codec.NewEncoderBytes(&bs, msgpackHandle)
+	err = enc.Encode(a)
+	if err != nil {
+		return nil, err
+	}
+
+	return bs, nil
+}
+
+func (e *ProxyConfigEntry) UnmarshalBinary(data []byte) error {
+	type alias ProxyConfigEntry
+
+	var a alias
+	dec := codec.NewDecoderBytes(data, msgpackHandle)
+	if err := dec.Decode(&a); err != nil {
+		return err
+	}
+
+	*e = ProxyConfigEntry(a)
+
+	config, err := lib.MapWalk(e.Config)
+	if err != nil {
+		return err
+	}
+
+	e.Config = config
+	return nil
+}
+
+func DecodeConfigEntry(raw map[string]interface{}) (ConfigEntry, error) {
+	var entry ConfigEntry
+
+	kindVal, ok := raw["Kind"]
+	if !ok {
+		kindVal, ok = raw["kind"]
+	}
+	if !ok {
+		return nil, fmt.Errorf("Payload does not contain a kind/Kind key at the top level")
+	}
+
+	if kindStr, ok := kindVal.(string); ok {
+		newEntry, err := MakeConfigEntry(kindStr, "")
+		if err != nil {
+			return nil, err
+		}
+		entry = newEntry
+	} else {
+		return nil, fmt.Errorf("Kind value in payload is not a string")
+	}
+
+	decodeConf := &mapstructure.DecoderConfig{
+		DecodeHook: mapstructure.StringToTimeDurationHookFunc(),
+		Result:     &entry,
+	}
+
+	decoder, err := mapstructure.NewDecoder(decodeConf)
+	if err != nil {
+		return nil, err
+	}
+
+	return entry, decoder.Decode(raw)
+}
```

Comment:
Is this used anywhere? I see (almost) the same func copy pasted into `api` package but missed any callers to this version. It confuses me because all the polymorphic marshalling in `structs` so far is based on custom encoding with kind prefixes so this is a little odd.

Reply:
The HTTP endpoint uses it.

It is quite convuluted process but basically when we are handling some JSON encoded data on the wire we have no idea what concrete structure to decode it into so all the struct marshalling cannot help us yet. In those cases we have to decode into a `map[string]interface{}` and then use this function to decode into a concrete type. Basically it looks for the `Kind` or `kind` in the map, generates a concrete type and then decodes into that. When encoding/decoding RPC request/responses we have the binary marshaller the kind will actually be encoded into the binary data twice. One up front to ensure we have a kind first so we can create a concrete type to decode the rest into and then within the config entry itself.

I thought a bit about whether the same `DecodeConfigEntry` function could be used in those as well. We could have a custom marshaller for a type alias of `[]ConfigEntry` which would decode first to a `[]map[string]interface{}` and then each one of those would have the `DecodeConfigEntry` func called on it. However that doesn't actually help the case where we are passing a single `ConfigEntry`. I don't believe we can define methods on an interface type. We could have another type alias for the `ConfigEntry` and define the marshalling interface on that but its just more indirection.

Really I dislike all of this polymorphic stuff and needing to do things in several different places. All I know is that what I have does actually produce the desired results. I would definitely be up for changing it I just haven't found any magic one way to do things which solves all the problems without the random binary marshaller implementations coupled with other supporting functions.

Reply:
It's called in the HTTP apply endpoint when decoding the incoming new config entry: https://github.com/hashicorp/consul/pull/5718/files#diff-aff7771a0694e9fbbc52e00c5ff45cbcR110

I think we need both methods - one is for encoding in RPCs with msgpack, the other is for the api when decoding from json.

Reply:
Yeah, I wanted to avoid the double decode pass through mapstructure in the msgpack case original which is why the custom binary for happened. I kinda overlooked the fact that we'd still need to do it when decoding from JSON.

Overall I think that's fine and this code is not especially unpleasant to achieve it. I don't see any red flags in it that will need constant changes when we add new types etc which was my largest concern with polymorphism in general.

I guess my only suggest is to maybe document this function to point out why it's needed and when it's appropriate (i.e. basically only when decoding from JSON, in general the msg pack marshalling takes care of decoding any internal representation.

Reply:
Comment added.